Composer/packages/client/src/components/TriggerCreationModal/TriggerOptionTree.ts

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { TriggerUISchema } from '@bfc/extension-client';
+
+import { TriggerUIOption } from './schema/TriggerOption';
+
+export class TriggerOptionLeafNode {
+  label: string;
+  $kind: string;
+  parent: TriggerOptionGroupNode | null = null;
+
+  constructor(label: string, $kind: string) {
+    this.label = label;
+    this.$kind = $kind;
+  }
+}
+
+export class TriggerOptionGroupNode {
+  label: string;
+  /** Title of a dropdown. 'Which activity type?' */
+  prompt?: string;
+  /** Placeholder of a dropdown input. 'Select an activity type' */
+  placeholder?: string;
+  children: (TriggerOptionLeafNode | TriggerOptionGroupNode)[] = [];
+  parent: TriggerOptionGroupNode | null = null;
+
+  constructor(label: string, prompt?: string, placeholder?: string) {
+    this.label = label;
+    this.prompt = prompt;
+    this.placeholder = placeholder;
+  }
+}
+
+export type TriggerOptionTree = TriggerOptionGroupNode;
+
+export type TriggerOptionTreeNode = TriggerOptionGroupNode | TriggerOptionLeafNode;
+
+const getGroupKey = (submenu) => (typeof submenu === 'object' ? submenu.label : submenu || '');
+
+export const generateTriggerOptionTree = (
+  triggerUIOptions: TriggerUISchema,
+  rootPrompt: string,
+  rootPlaceHolder: string,
+  optionCompareFn?: (a: TriggerOptionTreeNode, b: TriggerOptionTreeNode) => number
+): TriggerOptionTree => {
+  const root = new TriggerOptionGroupNode('triggerTypeDropDown', rootPrompt, rootPlaceHolder);
+
+  const leafNodeList = Object.entries(triggerUIOptions)
+    .filter(([, options]) => options && !options.submenu)
+    .map(([$kind, options]) => new TriggerOptionLeafNode(options?.label ?? '', $kind));
+  root.children.push(...leafNodeList);
+  leafNodeList.forEach((leaf) => (leaf.parent = root));
+
+  const groups = Object.values(triggerUIOptions)
+    .map((options) => options && options.submenu)
+    .filter((submenu) => !!submenu)
+    .reduce((result, submenu) => {
```

Comment:
This is probably better as a single `reduce` instead of the `map` -> `filter` -> `reduce`.

Reply:
made some improvements here